### ISSUE TITLE
Why is apiVersion and kind not set by default?

### ISSUE THREAD

**Issue by asteven** [gtsystem/lightkube#76]
Not really a problem, I'm just curious.

I would have expected the following to work:

```
In [1]: from lightkube import Client
   ...: from lightkube.resources.core_v1 import Pod
In [2]: p = Pod()
In [3]: assert p.kind == 'Pod'
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[3], line 1
----> 1 assert p.kind == 'Pod'

AssertionError: 

In [4]: assert p.apiVersion == 'v1'
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 assert p.apiVersion == 'v1'

AssertionError: 

In [5]: 
In [6]: assert Pod.apiVersion == 'v1'
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[6], line 1
----> 1 assert Pod.apiVersion == 'v1'

AssertionError:

In [7]: assert Pod.kind == 'Pod'
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[7], line 1
----> 1 assert Pod.kind == 'Pod'

AssertionError:

In [8]:

```

Information like apiVersion and kind is static and can never really be changed anyway, not?
I know the information is available in a resource as ._api_info, but why not just generate the models that way in the first place instead of defaulting to None?

For context: I'm writing a controller-runtime equivalent in python and it would have been useful to introspect that kind of information from a model class.

**Comment by asteven** [comment]
Possible implementation.
https://github.com/asteven/lightkube-models/tree/static-gvk

**Comment by gtsystem** [comment]
The data models are generated automatically from the openapi definition of the kubernetes API, and such fields have no default defined. See for example the definition of Pod:
```json
"io.k8s.api.core.v1.Pod": {
      "description": "Pod is a collection of containers that can run on a host. This resource is created by clients and scheduled onto hosts.",
      "properties": {
        "apiVersion": {
          "description": "APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources",
          "type": "string"
        },
        "kind": {
          "description": "Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds",
          "type": "string"
        },
  ... 
}
```
I'm not sure why this is the case, but I tried to respect their decision. Adding values for these attributes in the model now could be seen as a breaking change.

Regarding introspection, can you just use `isinstance(obj, Pod)`?
Alternatively the proper way to access the API info object is via the `api_info(..)` function defined in `lightkube.core.resource`.

**Comment by gtsystem** [comment]
> Still, it does not make any sense.

I fully agree, but this is the current status. Now, let's assume we would like to fix this

### Option 1 - Set `apiVersion` and `kind` as defaults in the model

After this change, a serialized model won't produce any longer these two attributes (defaults are not serialized). This could be ok when sending data in the wire, but it will break generation of YAML/JSON files with resource definitions where these attributes are mandatory. We could fix the included `dump_all_yaml` function but we still have few potential issues:
1. A user updating the model but not the client will still be affected.
2. Custom serialization functions will break

### Option 2 - Set `apiVersion` and `kind` with a ` __post_init__()`

With this solution, `apiVersion` and `kind` are set explictly and will always get serialized on the wire. This will increase the payload size but probably not by much, and they are anyway already set if resources are imported from YAML or if they got downloaded from the API server.

Last area of concern (which apply to both options) is if someone is explicity checking for the content of `apiVersion` and `kind`, expecting them to be unset. But I cannot think about any good reason to do that..

So between the two options I'll be more in favor of the second one

@ca-scribner, any thoughs or concerns on your side?

**Comment by asteven** [comment]
Hi there.

Thanks for coming back to this.

tl;dr: would it make sense for your project to be able to properly define Custom Resources and generate custom resource definitions from them? Then I'd be willing to discuss/contribute to that.


Let me give a little more context to where I'm coming from.

I want to write kubernetes operators in Python.

I started off with kopf, but that's basically dead and has a very abstract (and IMHO complex) code base which I fear only the original author is willing/capable of maintaining, but he's moved on.

So instead, I've ported the relevant parts of container-runtime to python, on top of anyio.
This is currently a PoC, but if it's survives a few smoketests will be open sourced after some initial code cleanup and possibly a migration to lightkube.

I started out with kubernetes-asyncio, which works ok'ish.
But since I'm dealing with many existing kubernetes resources, which all are static, and therefore can be typed, as well as custom resources, which the operator author has to define himself and are also static/typed I was looking for a typed solution instead of kubernetes-asyncio and found lightkube.

I currently use msgspec to define custom resources as Struct's and it's schema support allows me to generate CRD definitions which I can serialize to yaml/json and deploy them together with my operator. So I have the handling of custom resources solved.

This looks something like this:
```python
from msgspec import field
from untitled.resources import ObjectMeta, Resource, Spec, Status

class NaasVolumeClaimSpec(Spec):
    resources: GenericStruct = field(default_factory=GenericStruct)
    storageClassName: str = None
    volumeMode: str = None
    volumeName: str = None
    mountPath: str = None

class NaasVolumeClaimStatus(Status):
    pass

class NaasVolumeClaim(Resource):
    """NVMe-as-a-service volume claim."""
    __group__ = 'naas.xxx'
    __version__ = __version__
    __scope__ = 'Namespaced'
    __status_subresource__ = True
    __served__ = True
    __storage__ = True
    metadata: ObjectMeta
    spec: NaasVolumeClaimSpec
    status: NaasVolumeClaimStatus = field(default_factory=NaasVolumeClaimStatus)
```

ObjectMeta, Spec, Status and Resource are all msgspec.Struct subclasses.
Resource has a `__init_subclass__` method that registers resources with a in-mem registry, and `__post_init__` sets self.kind and self.apiVersion, similar/same as you mentioned in Option 2).

A small wrapper around `msgspec.json.schema` and `msgspec.yaml.encode` then produces yaml that can be fed into kubernetes as custom resource definitions.

But all upstream resources are just de-serialized as a 'GenericResource'.
That's where lightkube would come into play.

And if I could somehow generate the CRDs from lightkube's dataclasses, I could ditch msgspec and would have a full circle typed solution.

I can also solve this outside of lightkube. But maybe the definition of custom resources and the generation of CRD schemas could be things that make sense to have within lightkube.

**Comment by gtsystem** [comment]
> would it make sense for your project to be able to properly define Custom Resources and generate custom resource definitions from them? Then I'd be willing to discuss/contribute to that.

If I understand correctly, you would like to build custom resources and then automatically generate the corresponding CRD. 
However there are many CRDs already defined out there, and so it will be more valuable for the community a way to generate the models (custom resources) out of the CRD (following the API first approach).
No matter which direction you would like to take, it would be better this to be a separate repo/package so that it can be evolved at a faster pace. Of course, we can discuss of any change that would be needed in lightkube to support these toolings and I'll be glad to link these tools directly from the documentation.

Related issue gtsystem/lightkube-models#4

**Comment by asteven** [comment]
Just for reference, the controller runtime I mentioned is now public at: https://github.com/asteven/chopf

There's still a lot to do (docs, tests, types), but works for me - even using it in production ;-)
